packages/router/src/router.tsx

```diff
@@ -1,41 +1,221 @@
+/**
+ * External dependencies
+ */
+import RouteRecognizer from 'route-recognizer';
+import { createBrowserHistory } from 'history';
+
 /**
  * WordPress dependencies
  */
 import {
 	createContext,
 	useContext,
 	useSyncExternalStore,
+	useMemo,
 } from '@wordpress/element';
+import {
+	addQueryArgs,
+	getQueryArgs,
+	getPath,
+	buildQueryString,
+} from '@wordpress/url';
 
 /**
  * Internal dependencies
  */
-import history from './history';
-import type { EnhancedHistory } from './history';
+import type { ReactNode } from 'react';
+
+const history = createBrowserHistory();
+interface Route {
+	name: string;
+	path: string;
+	areas: Record< string, ReactNode >;
+	widths: Record< string, number >;
+}
 
-const RoutesContext = createContext< Location | null >( null );
-const HistoryContext = createContext< EnhancedHistory >( history );
+type LocationWithQuery = Location & {
+	query?: Record< string, any >;
+};
+
+interface Match {
+	name: string;
+	path: string;
+	areas: Record< string, ReactNode >;
+	widths: Record< string, number >;
+	query?: Record< string, any >;
+	params?: Record< string, any >;
+}
+
+export type BeforeNavigate = ( arg: {
+	path: string;
+	query: Record< string, any >;
+} ) => {
+	path: string;
+	query: Record< string, any >;
+};
+
+interface Config {
+	pathArg: string;
+	beforeNavigate?: BeforeNavigate;
+}
+
+export interface NavigationOptions {
+	transition?: string;
+	state?: Record< string, any >;
+}
+
+const RoutesContext = createContext< Match | null >( null );
+export const ConfigContext = createContext< Config >( { pathArg: 'p' } );
+
+const locationMemo = new WeakMap();
+function getLocationWithQuery() {
+	const location = history.location;
+	let locationWithQuery = locationMemo.get( location );
+	if ( ! locationWithQuery ) {
+		locationWithQuery = {
+			...location,
+			query: Object.fromEntries( new URLSearchParams( location.search ) ),
+		};
+		locationMemo.set( location, locationWithQuery );
+	}
+	return locationWithQuery;
+}
 
 export function useLocation() {
 	return useContext( RoutesContext );
 }
 
 export function useHistory() {
-	return useContext( HistoryContext );
+	const { pathArg, beforeNavigate } = useContext( ConfigContext );
```

Comment:
This could be a good case for the `useEvent` helper 🙂 The latest version of the callback is stored to a ref, `navigate` always calls the latest version, and you get rid of a `useMemo` dependency.

Reply:
Do you think we can get rid of the useMemo, isn't it better to keep the history object stable?

Reply:
Yes, the current minimal `useMemo` needs to remain there, because the `history` object returned from `useHistory()` is very often used as an `useEffect` or `useCallback` dependency.